pkg/task/storage.go

```diff
@@ -0,0 +1,155 @@
+package task
+
+import (
+	"container/heap"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+// TaskStorage is a leveldb-backed priority queue.
+// storage is persisted before the in-memory queue to prevent inconsistency between restarts; what
+// you see in the queue is the same as what is written the database
+type TaskStorage struct {
+	Max       int
+	Path      string
+	DBOpts    *opt.Options
+	WriteOpts *opt.WriteOptions
+	ReadOpts  *opt.ReadOptions
+	tq        *TaskQueue
+	db        *leveldb.DB
+	mux       sync.Mutex
+}
+
+// Open database and load its contents into memory.
+func (s *TaskStorage) Open() error {
+	db, err := leveldb.OpenFile(s.Path, s.DBOpts)
+	if err != nil {
+		return err
+	}
+	s.db = db
+	return s.Reload()
+}
+
+// Method for closing the database. Always run Close when finished.
+func (s *TaskStorage) Close() error {
+	s.tq = nil
+	return s.db.Close()
+}
+
+// Read everything from the database into memory. Typically, you will not need to run this.
+func (s *TaskStorage) Reload() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.tq = new(TaskQueue)
+	iter := s.db.NewIterator(nil, s.ReadOpts)
+	for iter.Next() {
+		tsk := new(Task)
+		err := json.Unmarshal(iter.Value(), tsk)
+		if err != nil {
+			return err
+		}
+		if tsk.State != TaskStateComplete {
+			heap.Push(s.tq, tsk)
+		}
+	}
+	iter.Release()
+	return iter.Error()
+}
+
+// Add an item to the priority queue
+func (s *TaskStorage) Push(tsk *Task) error {
```

Comment:
I find this name a bit confusing - `TaskStorage` sounds like the thing that takes care of storage for `Tasks`, but actually this type has the semantics for managing the state as well as storing - we modify `tsk.State` on Push, check for `s.Max`, etc.

Reply:
Maybe just `Queue` is a better name for this type, then the current `TaskQueue` doesn't need to be exported, it could be a private type AFAICT. Developers would have to interact only with this `Queue` I think, and with the `Task` type.

Reply:
Good point. Unexported TaskQueue and changed the name of TaskStorage to Queue.